### PR TITLE
Redesign anchor creation

### PR DESCRIPTION
**Pull request by fenjalien**
Elements now have to always create their own anchors instead of it happening in `process` by inferring the element's bounds. Elements must now return a function that takes a string and outputs the transformed anchor. An empty array can be passed to the function to get the list of anchor names.

I've only made these changes to `circle` as I want feedback before I go ahead. Its not perfect and could do with some tidying up as well.

I think this can also pave the way for anchor along a path and angle anchors, as they both would require a callback.

### CODE REVIEW

**Comment by johannes-wolf** [comment]
I am sorry, I think that I have introduced all those merge conflicts by updating the `redesign-internals` branch on the current master, as it was missing one commit, unaware that this would cause problems.

**Comment by fenjalien** [comment]
@johannes-wolf what do you want to do about bounding box anchors for groups? They've been removed from shapes so far.

**Comment by johannes-wolf** [comment]
We shuold add them Getting a bbox of multiple elements is very important.

**Comment by fenjalien** [comment]
Current notes:
- Arc angles cannot be `0deg` as they rely on border anchors for caridnal anchors. So by allowing `0deg` they have no volume/essentially a straight line which then can't be tested for intersections...
- I broke content rotation and idk how to fix it

**Comment by johannes-wolf** [comment]
> Current notes:
> 
>     * Arc angles cannot be `0deg` as they rely on border anchors for caridnal anchors. So by allowing `0deg` they have no volume/essentially a straight line which then can't be tested for intersections...

Wdym with angles 0deg? Stop-Start != 0deg? I think that is fine.

>     * I broke content rotation and idk how to fix it

I fixed the rotation of content.

Some tests fail with panics.

**Comment by johannes-wolf** [comment]
@fenjalien I fixed many things but group translation, which do not work at all.

**Review by johannes-wolf — CHANGES_REQUESTED** [review]
Group translation (`anchor: "..."`) currently does nothing.

File: src/anchor.typ [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+//     util.apply-transform(
+//       transform,
+//       out
+//     )
+//   } else {
+//     out
+//   }
+// }
+
+#let setup(callback, anchor-names, default: none, transform: none, name: none, offset-anchor: none) = {
```
Comment (johannes-wolf) [inline src/anchor.typ:33]: I have to say, I do not really get this function.
How can I do calculation on (transformed) coordinates for anchor creation?
Example: Line midpoint (which got removed, why?).

File: src/anchor.typ [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        out
+      )
+    } else {
+      out
+    }
+  }
+  return (transform, calculate-anchor)
+}
+
+#let border(center, x-dist, y-dist, drawables, angle) = {
```
Comment (johannes-wolf) [inline src/anchor.typ:74]: Suggest renaming this into `frame`.

File: src/anchor.typ (hunk 1 above)
Comment (fenjalien) [reply src/anchor.typ:46]: I've added some docs to this but it mainly returns another function that can be called to get the position of an anchor. Is there a situation where you would need to use transformed coordinates for anchor creation? Even so you can transform the anchors before and leave `transfrom` as `none`.

Also line never had a midpoint anchor? do you mean `center`? which it really shouldn't have. `mid` can be added with path anchors though!

File: src/anchor.typ (hunk 2 above)
Comment (fenjalien) [reply src/anchor.typ:103]: I suggest not as tikz refers to them as *border* anchors and this function should be used to calculate them.

**Comment by fenjalien** [comment]
@johannes-wolf fixed the group anchor issue and several others. I would like to add some tests but I'm not sure what exactly.

**Comment by johannes-wolf** [comment]
Some things to test:
- Group anchors with and without inner and outer transformations
- Group translation
- Anchors for all the draw commands

File: src/anchor.typ (hunk 1 above)
Comment (johannes-wolf) [reply src/anchor.typ:46]: Ok. Can you give an example how one could implement a custom anchor that gets calculated in the anchor callback? Something like `mid`?

**Comment by johannes-wolf** [comment]
Rebased onto 0.2.0.

File: src/coordinate.typ [hunk 3]
```diff
@@ -68,16 +68,22 @@
...
-    message: strfmt("Unknown element '{}' in elements {}", name, repr(ctx.nodes.keys()))
-  )
+  // Two stage anchor lookup:
+  //   1: elements => <anchor>
+  //   2: current group => anchors => <anchor>
+  let anchor = if name in ctx.nodes {
+    let node = ctx.nodes.at(name)
+    (node.anchors)(anchor)
+  } else if name in ctx.groups.last().anchors {
+    ctx.groups.last().anchors.at(name)
```
Comment (johannes-wolf) [inline src/coordinate.typ:78]: Not sure about this: This way, anchors added via `copy-anchors` can not overwrite element anchors.

**Pull request merged**